test/foundry/utils/version-control/RoninValidatorSetTimedMigrator.sol

```diff
@@ -0,0 +1,175 @@
+import { ILogicValidatorSet, MockLogicValidatorSetV1, MockLogicValidatorSetV2 } from "@ronin/contracts/mocks/utils/version-control/MockLogicValidatorSet.sol";
+import { MockActor } from "@ronin/contracts/mocks/utils/version-control/MockActor.sol";
+import { RoninValidatorSetTimedMigrator } from "@ronin/contracts/utils/version-control/RoninValidatorSetTimedMigrator.sol";
+
+contract RoninValidatorSetTimedMigratorTest is Test {
+  /**
+   * @dev Error thrown when a duplicated element is detected in an array.
+   * @param msgSig The function signature that invoke the error.
+   */
+  error ErrDuplicated(bytes4 msgSig);
+  /// @dev Error of set to non-contract.
+  error ErrZeroCodeContract(address addr);
+  /// @dev Error when contract which delegate to this contract is not compatible with ERC1967
+  error ErrDelegateFromUnknownOrigin(address addr);
+  /**
+   * @dev Error indicating that a function can only be called by the contract itself.
+   * @param msgSig The function signature (bytes4) that can only be called by the contract itself.
+   */
+  error ErrOnlySelfCall(bytes4 msgSig);
+
+  address logicV1;
+  address logicV2;
+
+  address payable proxyStorage;
+  address payable versionController;
+
+  address admin;
+  address alice;
+  MockActor bobContract;
+
+  /**
+   * @dev Emitted when the implementation is upgraded.
+   */
+  event Upgraded(address indexed implementation);
+
+  function setUp() public {
+    admin = vm.addr(555);
+    alice = vm.addr(1);
+
+    logicV1 = address(new MockLogicValidatorSetV1());
+    logicV2 = address(new MockLogicValidatorSetV2());
+
+    vm.prank(admin, admin);
+
+    proxyStorage = payable(address(new TransparentUpgradeableProxyV2(logicV1, admin, "")));
+    versionController = payable(address(new RoninValidatorSetTimedMigrator(proxyStorage, logicV1, logicV2)));
+    bobContract = new MockActor(proxyStorage);
+  }
+
+  function testManualUpgradeToVersionControl() public {
+    _testManualUpgrade(versionController);
+  }
+
+  function _testManualUpgrade(address implementation) private {
+    vm.expectEmit(true, false, false, false);
+    emit Upgraded(implementation);
+    vm.startPrank(admin, admin);
+    TransparentUpgradeableProxyV2(proxyStorage).upgradeTo(implementation);
+    assertEq(implementation, TransparentUpgradeableProxyV2(proxyStorage).implementation());
+    vm.stopPrank();
+  }
+
+  function testFailDupplicateInput0() public {
+    vm.expectRevert(abi.encodePacked(ErrDuplicated.selector, bytes4(0)));
+    new RoninValidatorSetTimedMigrator(proxyStorage, proxyStorage, proxyStorage);
+  }
+
+  function testFailDupplicateInput1() public {
+    vm.expectRevert(abi.encodePacked(ErrDuplicated.selector, bytes4(0)));
+    new RoninValidatorSetTimedMigrator(proxyStorage, logicV1, logicV1);
+  }
+
+  function testFailDupplicateInput2() public {
+    vm.expectRevert(abi.encodePacked(ErrDuplicated.selector, bytes4(0)));
+    new RoninValidatorSetTimedMigrator(logicV1, proxyStorage, logicV1);
+  }
```

Comment:
Not so support this approach, consider keeping one of them and adding fuzz test for this case, ie:
```solidity
  function testFailDupplicateInput(uint8[3] memory dupIds) public {
    assume(dupIds[i] < 3);
    address[3] memory addrs = _constructAddressesFromIdxs(dupIds);
    new RoninValidatorSetTimedMigrator(logicV1, proxyStorage, logicV1);
  }
```